plugins/GSdx/GSTextureCache.cpp

```diff
@@ -1598,9 +1585,12 @@ GSTextureCache::Source::Source(GSRenderer* r, const GIFRegTEX0& TEX0, const GIFR
 
 GSTextureCache::Source::~Source()
 {
-	m_renderer->m_dev->Recycle(m_palette);
-
-	_aligned_free(m_clut);
+	if (!m_should_have_tex_palette) {
```

Comment:
I don't like this solution. Desctruction should be done in Palette object. I don't know what will be the speed impact but maybe you could use `shared_pointer<T>` instead. That being said, we should delete palette only when there are unused from a long time (potentially palette will be reused next frame)

Reply:
The code path which recycles palette on Source destruction is potentially hit only by ICO (palette allocated outside the `GSTextureCache` scope when 32 bit textures, so `m_should_have_tex_palette` is false), but if this is not the case when the `PaletteMap` is not used, then `m_palette` in Source is always null and no recycling outside `Palette` destruction happens.
Desctruction is already done in `PaletteMap` in fact for all the normal cases, and only on `Clear`, which is called when `RemoveAll` is called in the `GSTextureCache`.

Reply:
It feels way too complex for me. ICO should be updated to a GetPallette (or SetPalette). It should work as the texture cache.

Potentially it could be interesting to use some kind of shared_ptr to handle the reference counting of allocation.

Reply:
Can a better detection for ICO be added, one that doesn't rely on crc lookup ?